File: GroceryListHelper.Client/Services/CartProductsApiService.cs
namespace GroceryListHelper.Client.Services;

public sealed class CartProductsApiService(IHttpClientFactory clientFactory) : ICartProductsService
{
    private readonly HttpClient client = clientFactory.CreateClient("Client");
    private const string uri = "api/cartproducts";

    public async Task<List<CartProductCollectable>> GetCartProducts()
    {
        return await client.GetFromJsonAsync<List<CartProductCollectable>>(uri) ?? [];
    }

    public async Task CreateCartProduct(CartProduct product)
    {
        HttpResponseMessage response = await client.PostAsJsonAsync(uri, product);
        response.EnsureSuccessStatusCode();
    }

    public async Task DeleteCartProduct(string productName)
    {
        HttpResponseMessage response = await client.DeleteAsync(uri + $"/{productName}");
        response.EnsureSuccessStatusCode();
    }

    public async Task DeleteAllCartProducts()
    {
        await client.DeleteAsync(uri);
    }

    public async Task UpdateCartProduct(CartProductCollectable cartProduct)
    {
        await client.PutAsJsonAsync(uri, cartProduct);
    }

    public async Task SortCartProducts(ListSortDirection sortDirection)
    {
        await client.PatchAsync(uri + "/sort/" + (int)sortDirection, null);
    }
}

File: GroceryListHelper.Client/src/services/CartProductsApiService.ts
import type { CartProduct } from "../types/CartProduct";
import type { ICartProductsService } from "../types/ICartProductsService";
import type { SortDirection } from "../types/SortState";
import { send } from "../helpers/httpHelpers";

export class CartProductsApiService implements ICartProductsService {
    readonly url = "api/cartproducts/";

    getCartProducts = async () => {
        const productsResponse = await fetch(this.url);
        const products = await productsResponse.json() as CartProduct[];
        return products;
    }

    createCartProduct = async (product: CartProduct) => {
        const response = await send(this.url, "POST", product);
        return response;
    }

    updateCartProduct = async (product: CartProduct) => {
        const response = await send(this.url, "POST", product);
        return response;
    }

    sortCartProducts = async (sortDirection: SortDirection) => {
        const url = this.url + "sort/" + (sortDirection === "Ascending" ? "0" : "1");
        const response = await send(url, "PATCH");
        return response;
    }

    deleteAllCartProducts = async () => {
        const response = await send(this.url, "DELETE");
        return response;
    }

    deleteCartProduct = async (name: string) => {
        const response = await send(this.url + name, "DELETE");
        return response;
    }
}